lib/limiter/reservation_test.go

```diff
@@ -0,0 +1,229 @@
+	var rsvr ClientReserver = NewUniformlyBoundedClientReserver(maxReservationsPerClient)
+
+	alice := DummyClientID("alice")
+	bob := DummyClientID("bob")
+	ctx := context.Background()
+
+	rb1, err := rsvr.TryReserve(ctx, bob)
+	require.NoError(t, err)
+
+	rb2, err := rsvr.TryReserve(ctx, bob)
+	require.NoError(t, err)
+
+	ra1, err := rsvr.TryReserve(ctx, alice)
+	require.NoError(t, err)
+
+	err = rsvr.ReleaseReservation(ctx, rb1)
+	require.NoError(t, err)
+
+	ra2, err := rsvr.TryReserve(ctx, alice)
+	require.NoError(t, err)
+
+	rb3, err := rsvr.TryReserve(ctx, bob)
+	require.NoError(t, err)
+
+	ra3, err := rsvr.TryReserve(ctx, alice)
+	require.Equal(t, MaxReservationsExceeded, err)
+	require.Equal(t, ClientReservation{}, ra3)
+
+	err = rsvr.ReleaseReservation(ctx, ra2)
+	require.NoError(t, err)
+
+	rb4, err := rsvr.TryReserve(ctx, bob)
+	require.Equal(t, MaxReservationsExceeded, err)
+	require.Equal(t, ClientReservation{}, rb4)
+
+	err = rsvr.ReleaseReservation(ctx, ra1)
+	require.NoError(t, err)
+
+	err = rsvr.ReleaseReservation(ctx, rb2)
+	require.NoError(t, err)
+
+	err = rsvr.ReleaseReservation(ctx, rb3)
+	require.NoError(t, err)
+
+	requireAllCountsZero(t, rsvr.(*UniformlyBoundedClientReserver))
+}
+
+func TestUniformlyBoundedClientReserverConcurrent(t *testing.T) {
+	// Randomised scenario of concurrent reservation attempts by two clients.
+	// The intent of this test is to potentially identify data races.
+
+	// We start a number of worker threads for two clients. Each worker thread
+	// iteratively attempts to reserve, pause, then release a reservation.
+
+	// BEWARE: if the reserver implementation is defective, the behaviour
+	// of this test may be nondeterministic as it depends upon goroutine
+	// scheduling.
+
+	// Pin random seed. Used to control worker thread pauses between
+	// successive reservation attempts.
+	seed := int64(12345)
+	source := rand.NewSource(seed)
+	random := rand.New(source)
+
+	// create a reserver with some per client limit
+	var maxReservationsPerClient int64 = 5
+	var rsvr ClientReserver = NewUniformlyBoundedClientReserver(maxReservationsPerClient)
+
+	// two clients, alice and bob
+	alice := DummyClientID("alice")
+	bob := DummyClientID("bob")
+
+	clients := []DummyClientID{alice, bob}
+
+	type workerStats struct {
+		Client        core.ClientID
+		Reserved      int64
+		Limited       int64
+		Errors        int64
+		ReleaseErrors int64
+	}
+
+	// Create twice as many workers per client as the limit
+	// init each worker with same max iters and a different seed
+	var wg sync.WaitGroup
+
+	workersPerClient := 2 * maxReservationsPerClient
+	itersPerWorker := int64(1000)
+
+	stats := make(chan workerStats, int64(len(clients))*workersPerClient)
+
+	worker := func(c core.ClientID, iters int64, seed int64, rsvr ClientReserver, out chan<- workerStats) {
+		defer wg.Done()
+		var s workerStats
+		s.Client = c
+		ctx := context.Background()
+
+		source := rand.NewSource(seed)
+		random := rand.New(source)
```

Comment:
source and random redeclared? Reuse the top-level ones (or remove them)?

Reply:
the docs state that non default sources aren't safe for use by multiple goroutines, so i created a new one per worker.  but your excellent suggestion to remove the randomised sleep completely also solves this issue